nmmo/core/observation.py

```diff
@@ -100,15 +118,193 @@ def to_gym(self):
     if self.config.ITEM_SYSTEM_ENABLED:
       gym_obs["Inventory"] = np.vstack([
         self.inventory.values, np.zeros((
-          self.config.ITEM_N_OBS - self.inventory.values.shape[0],
+          self.config.INVENTORY_N_OBS - self.inventory.values.shape[0],
           self.inventory.values.shape[1]))
       ])
 
     if self.config.EXCHANGE_SYSTEM_ENABLED:
       gym_obs["Market"] = np.vstack([
         self.market.values, np.zeros((
-          self.config.EXCHANGE_N_OBS - self.market.values.shape[0],
+          self.config.MARKET_N_OBS - self.market.values.shape[0],
           self.market.values.shape[1]))
       ])
 
+    gym_obs["ActionTargets"] = self.generate_action_targets()
+
     return gym_obs
+
+  def generate_action_targets(self):
+    # TODO(kywch): return all-0 masks for buy/sell/give during combat
+
+    masks = {}
+    masks[action.Move] = {
+      action.Direction: self._generate_move_mask()
+    }
+
+    if self.config.COMBAT_SYSTEM_ENABLED:
+      masks[action.Attack] = {
+        action.Style: self._generate_allow_all_mask(action.Style.edges),
+        action.Target: self._generate_attack_mask()
+      }
+
+    if self.config.ITEM_SYSTEM_ENABLED:
+      masks[action.Use] = {
+        action.InventoryItem: self._generate_use_mask()
+      }
+
+    if self.config.EXCHANGE_SYSTEM_ENABLED:
+      masks[action.Sell] = {
+        action.InventoryItem: self._generate_sell_mask(),
+        action.Price: None # allow any integer
+      }
+      masks[action.Buy] = {
+        action.MarketItem: self._generate_buy_mask()
+      }
+
+    if self.config.COMMUNICATION_SYSTEM_ENABLED:
+      masks[action.Comm] = {
+        action.Token: self._generate_allow_all_mask(action.Token.edges),
+      }
+
+    return masks
+
+  def _generate_allow_all_mask(self, actions):
+    return np.ones(len(actions), dtype=np.int8)
+
+  def _generate_move_mask(self):
+    # pylint: disable=not-an-iterable
+    return np.array(
+      [self.tile(*d.delta).material_id in material.Habitable
+       for d in action.Direction.edges], dtype=np.int8)
+
+  def _generate_attack_mask(self):
+    # TODO: Currently, all attacks have the same range
+    #   if we choose to make ranges different, the masks
+    #   should be differently generated by attack styles
+    assert self.config.COMBAT_MELEE_REACH == self.config.COMBAT_RANGE_REACH
+    assert self.config.COMBAT_MELEE_REACH == self.config.COMBAT_MAGE_REACH
+    assert self.config.COMBAT_RANGE_REACH == self.config.COMBAT_RANGE_REACH
+
+    attack_range = self.config.COMBAT_MELEE_REACH
+
+    agent = self.agent()
+    dist_from_self = self.entities.dist((agent.row, agent.col))
+    not_same_tile = dist_from_self > 0 # this also includes not_self
```

Comment:
you can attack enemies in the same tile.

Reply:
removed!